src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -582,19 +596,96 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+    {
+        Debug.Assert(factoryContext.RequestBodyParameter is not null, "factoryContext.RequestBodyParameter is null for a body parameter.");
+        Debug.Assert(factoryContext.RequestBodyParameter.Name is not null, "CreateArgument() should throw if parameter.Name is null.");
+
+        if (factoryContext.ParameterBinders.Count > 0)
+        {
+            // We need to generate the code for reading from the body before calling into the delegate
+            var continuation = Expression.Lambda<Func<object?, HttpContext, ReadOnlySequence<byte>, object?[], Task>>(
+            responseWritingMethodCall, TargetExpr, HttpContextExpr, RawBodyValueExpr, BoundValuesArrayExpr).Compile();
+
+            // Looping over arrays is faster
+            var binders = factoryContext.ParameterBinders.ToArray();
+            var count = binders.Length;
+
+            return async (target, httpContext) =>
+            {
+                // Run these first so that they can potentially read and rewind the body
+                var boundValues = new object?[count];
+
+                for (var i = 0; i < count; i++)
+                {
+                    boundValues[i] = await binders[i](httpContext);
+                }
+
+                var bodyValue = await ReadBodyAsync(httpContext);
+
+                await continuation(target, httpContext, bodyValue, boundValues);
+            };
+        }
         else
         {
-            return HandleRequestBodyAndCompileRequestDelegateForJson(responseWritingMethodCall, factoryContext);
+            // We need to generate the code for reading from the body before calling into the delegate
+            var continuation = Expression.Lambda<Func<object?, HttpContext, ReadOnlySequence<byte>, Task>>(
+            responseWritingMethodCall, TargetExpr, HttpContextExpr, RawBodyValueExpr).Compile();
+
+            return async (target, httpContext) =>
+            {
+                var bodyValue = await ReadBodyAsync(httpContext);
+
+                await continuation(target, httpContext, bodyValue);
+            };
+        }
+
+        static async ValueTask<ReadOnlySequence<byte>> ReadBodyAsync(HttpContext httpContext)
+        {
+            var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
+
+            if (feature?.CanHaveBody == true)
+            {
+                var bodyReader = httpContext.Request.BodyReader;
+
+                while (true)
+                {
+                    var result = await bodyReader.ReadAsync();
+                    var buffer = result.Buffer;
+
+                    if (result.IsCompleted)
+                    {
+                        // We're not buffering the body so we want to block consuming code from reading again
+                        // and getting weird errors. Treat further reads as a fully consumed body.
+                        httpContext.Request.Body = Stream.Null;
```

Comment:
This is odd, what if someone else is buffering the body?

Reply:
I guess you'd have to check if the body is seekeable, if so seek it. RadFormAsync / MVC doesn't change / reset the Body once it has been read, couldn't we do the same here?

Reply:
I think seeking given a seekable stream is fine.

If we're replacing the body anyway, we should advance the reader before returning even if Kestrel will do it anyway.

We would have to do this after invoking the route handler, so memory doesn't get returned too soon. I'm still worried about people keeping a reference to the `ReadOnlySequence<byte>` after the route handler completes if we don't copy.

Reply:
> If we're replacing the body anyway, we should advance the reader before returning even if Kestrel will do it anyway.

Just for cleanliness? It's not just Kestrel doing it, it would happen for any properly implemented pipereader (it will complete the pipereader).

> I'm still worried about people keeping a reference to the ReadOnlySequence<byte> after the route handler completes if we don't copy.

Yea but this feature is about not copying the buffer. That's why it's a `ROS<byte>`. That's the one thing I'm unwilling to compromise here on as I think it makes the original scenario (posted in the issue) less efficient.

Reply:
> This is odd, what if someone else is buffering the body?

I don't see how that affects this? I can revert and set the the original stream after executing the handler. The body is being buffered for the logic that is consuming the buffer.

Reply:
> Just for cleanliness? It's not just Kestrel doing it, it would happen for any properly implemented pipereader (it will complete the pipereader).

There's no guarantee the next call will be to `Complete()`. It could matter if someone writes request body draining middleware that waits a different amount of time than Kestrel's 5 seconds. Basically, why not? Especially if we end up doing Stream resetting logic after the route handler completes anyway.

Reply:
We have the reference to the pipe that we read and we can complete it. It's just redundant is all.

Reply:
It's more about generic code that might try to read any unconsumed bytes from the body after the route handler runs but before middleware completes.

Reply:
Also @Tratcher I don't think we should check is seekable here. Specifically because we need to poison both the stream and the PipeReader to be unusable 😄

Reply:
At least revert it when you're done then.

Reply:
Done.